### PR TITLE
test Base Directory should be created if not exist

### PR DESCRIPTION
**Pull request by guangxuli**
fix https://github.com/openshift/origin/issues/13367

### CODE REVIEW

**Comment by smarterclayton** [comment]
@bparees looks like you also made this change?

**Comment by bparees** [comment]
(and I don't think checking for the existence of the dir is necessary, mkdirall will silently return if the dir exists anyway)

**Comment by bparees** [comment]
@guangxuli unless you have strong objections, i'm going to go w/ my fix here: https://github.com/openshift/origin/pull/13397

**Comment by guangxuli** [comment]
@bparees I attention is that if the base dir is created by the test case, we should remove the test stuff created by this unit test case, and if not , we should not remove the base dir. But i found my previous fix also have some problem that when the base dir is not create by the test case, I remove it also. That's incorrect. I have fix that point.

File: test/util/client.go [hunk 1]
```diff
@@ -31,6 +31,20 @@ func GetBaseDir() string {
...
+	if _, err = os.Stat(baseDir); err == nil {
+		return false, nil
+	}
+	if os.IsNotExist(err) {
+		if err = os.MkdirAll(baseDir, os.FileMode(0766)); err != nil {
+			return false, err
+		}
+	}
+	return true, nil
+}
```
Comment (bparees) [inline test/util/client.go]: i still don't think this is necessary.  just do mkdirall, it no-ops if the dir exists.

Comment (bparees) [inline test/util/client.go]: hm, ok i see you're using the determination to setup the defer.

**Comment by bparees** [comment]
lgtm
[merge]

**Comment by bparees** [comment]
looks like basedir is set to nil when this is run in certain environments:

```
=== RUN   TestGetDockerfileFrom
--- FAIL: TestGetDockerfileFrom (0.00s)
	docker_test.go:351: failed to create BaseDir: <nil>
```

File: pkg/build/builder/docker_test.go [hunk 2]
```diff
@@ -338,6 +344,13 @@ RUN echo "hello world"
 			want: []string{"scratch", "busybox"},
 		},
 	}
+
+	if isNew, err := util.CreateBaseDir(util.GetBaseDir()); err == nil && isNew {
+		defer os.RemoveAll(util.GetBaseDir())
+	} else {
+		t.Errorf("failed to create BaseDir: %v", err)
+	}
```
Comment (bparees) [inline pkg/build/builder/docker_test.go]: this needs to be reworked like the other one so you don't error out if isNew==false.

Comment (guangxuli) [inline pkg/build/builder/docker_test.go]: yes, done

**Comment by guangxuli** [comment]
@bparees PTAL

**Comment by bparees** [comment]
[merge]

File: pkg/build/builder/docker_test.go [hunk 3]
```diff
@@ -192,6 +192,14 @@ func TestDockerfilePath(t *testing.T) {
 		"\"io.openshift.build.commit.message\"=\"message\"",
 	}
 
+	if isNew, err := util.CreateBaseDir(util.GetBaseDir()); err == nil {
```
Comment (smarterclayton) [inline pkg/build/builder/docker_test.go]: Unit tests should not be creating output into BASETMPDIR - they should simply use the go tempdir function

Comment (guangxuli) [inline pkg/build/builder/docker_test.go]: thanks @smarterclayton, indeed, i should simply use` ioutil.TempDir("", "dockerfile-path")` @bparees sorry, how do you think, should i create a new pull request ?

Comment (bparees) [inline pkg/build/builder/docker_test.go]: sure.

**Pull request merged**